### PR TITLE
Merge ifcfgs from AY and linuxrc when keep_install_network is set

### PR DESCRIPTION
**Pull request by mchf**
bnc#874259

Porting into SLE-12 for OES team.

Not ready for merging, I just want to see if Travis is happy.

### CODE REVIEW

**Comment by mvidner** [comment]
For cross reference, this is a port of #296 (SLE11-SP3) and #297 (SLE11-SP4).

File: src/lib/network/network_autoyast.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    # wins.
+    #
+    # @param [Hash] in_devs1 first map of devices in NetworkInterfaces format
+    # @param [Hash] in_devs2 second map of devices in NetworkInterfaces format
+    #
+    # @return merged device map in NetworkInterfaces format or empty map
+    def merge_devices(in_devs1, in_devs2)
+      return in_devs2 if in_devs1.nil? && !in_devs2.nil?
+      return in_devs1 if in_devs2.nil? && !in_devs1.nil?
+      return {} if in_devs1.nil? && in_devs2.nil?
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: The three complex conditions can be replaced with:
```ruby
in_devs1 ||= {}
in_devs2 ||= {}
```

I know I am repeating myself, and that you prefer early returns. But the code complexity weighs much against early returns in this case, it seems.

File: src/lib/network/network_autoyast.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    #     ...
+    #   },
+    #   ...
+    # }
+    #
+    # If a device definition is present in both maps, then the one from devices2
+    # wins.
+    #
+    # @param [Hash] in_devs1 first map of devices in NetworkInterfaces format
+    # @param [Hash] in_devs2 second map of devices in NetworkInterfaces format
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: `[Hash, nil]`, see http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md#Declaring_Types

File: src/clients/lan_auto.rb [hunk 3]
```diff
@@ -96,37 +92,19 @@ def main
...
           Lan.Read(:cache)
           # export settings into AY map
           @from_system = Lan.Export
-          @dns = Ops.get_map(@from_system, "dns", {})
-          @routing = Ops.get_map(@from_system, "routing", {})
+          @dns = @from_system["dns"] || {}
+          @routing = @from_system["routing"] || {}
 
-          # copy the keys/values that are not existing in the XML
+          # copy the keys/values that are not existing in the XMomm
```
Comment (mvidner) [inline src/clients/lan_auto.rb]: xmwtf?

File: src/clients/lan_auto.rb [hunk 4]
```diff
@@ -96,37 +92,19 @@ def main
...
-                "(dns) taking %1 from inst-sys. Value = %2",
-                key,
-                value
-              )
-              Ops.set(@param, ["dns", key], value)
-            end
-          end
+          @param["dns"] = NetworkAutoYast.instance.merge_dns(@param["dns"] || {}, @dns)
+
           Ops.set(@param, "routing", {}) if !Builtins.haskey(@param, "routing")
```
Comment (mvidner) [inline src/clients/lan_auto.rb]: If we're touching `merge_dns` then we should touch the next chunk of code which does a very similar thing.

File: src/lib/network/network_autoyast.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    #
+    # bnc#796580 The problem with this is that due to compatibility with
+    # older profiles, a missing element may have a different meaning than
+    # "use what the filesystem/kernel currently uses".
+    # In particular, a missing write_hostname means
+    # "use the product default from DVD1/control.xml".
+    # Other elements may have similar problems,
+    # to be fixed post-PTF for maintenance.
+    def merge_dns(ay_dns, instsys_dns)
+      raise ArgumentError if ay_dns.nil? || instsys_dns.nil?
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: It seems inconsistent to start a new class and let it allow nils in one method and disallow them in another method. Maybe `merge_devices` should require the *caller* handle them: 
```
NetworkAutoYast.instance.merge_devices(input["devices"] || {}, devices || {})
```

File: src/lib/network/network_autoyast.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    # case of key collision.
+    #
+    # bnc#796580 The problem with this is that due to compatibility with
+    # older profiles, a missing element may have a different meaning than
+    # "use what the filesystem/kernel currently uses".
+    # In particular, a missing write_hostname means
+    # "use the product default from DVD1/control.xml".
+    # Other elements may have similar problems,
+    # to be fixed post-PTF for maintenance.
+    def merge_dns(ay_dns, instsys_dns)
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]:
```
# @param ay_dns [Hash]
# @param instsys_dns [Hash]
```
Also, the order of the arguments should be reversed to match [`Hash#merge`](http://ruby-doc.org/core-2.1.0/Hash.html#method-i-merge): `loser.merge(winner)`

File: src/clients/lan_auto.rb [hunk 7]
```diff
@@ -96,49 +92,8 @@ def main
...
-              Ops.set(@param, ["routing", key], value)
-              Builtins.y2milestone(
-                "(routing) taking %1 from inst-sys. Value = %2",
-                key,
-                value
-              )
-            end
-          end
-        end
+        @param = NetworkAutoYast.instance.merge_configs if @param["keep_install_network"]
```
Comment (mvidner) [inline src/clients/lan_auto.rb]: `merge_configs(@param)`

**Comment by mvidner** [comment]
Just one issue, but an important one :)

File: src/clients/lan_auto.rb (hunk 7 above)
Comment (mchf) [inline src/clients/lan_auto.rb]: ah sure, thank you

File: src/lib/network/network_autoyast.rb (hunk 1 above)
Comment (kobliha) [inline src/lib/network/network_autoyast.rb]: @mvidner Is this blocking the merge?

**Comment by kobliha** [comment]
I want to see this merged ASAP, is blocks yet another "story" to be finished.

**Comment by mvidner** [comment]
LGTM!

**Comment by mchf** [comment]
Thanks for review

**Pull request merged**